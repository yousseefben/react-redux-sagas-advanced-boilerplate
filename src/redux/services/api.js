export function request(url, options = {}) {
  const config = {
    method: 'GET',
    ...options
  };
  const errors = [];

  if (!url) {
    errors.push('url');
  }

  if (
    !config.payload &&
    (config.method !== 'GET' && config.method !== 'DELETE')
  ) {
    errors.push('payload');
  }

  if (errors.length) {
    throw new Error(`Error! You must pass \`${errors.join('`, `')}\``);
  }

  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...config.headers
  };

  const params = {
    headers,
    method: config.method
  };

  if (params.method !== 'GET') {
    params.body = JSON.stringify(config.payload);
  }

  return fetch(url, params).then(async response => {
    const contentType = response.headers.get('content-type');

    if (response.status > 299) {
      const error = {};
      error.status = response.status;

      if (contentType && contentType.includes('application/json')) {
        error.response = await response.json();
      } else {
        error.response = await response.text();
      }

      throw error;
    } else {
      if (contentType && contentType.includes('application/json')) {
        return response.json();
      }

      return response.text();
    }
  });
}
