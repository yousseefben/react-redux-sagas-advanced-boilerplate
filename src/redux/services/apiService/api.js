import axios from 'axios';

export function request(url, options = {}) {
  const config = {
    method: 'GET',
    ...options
  };
  const errors = [];

  if (!url) {
    errors.push('url');
  }

  if (
    !config.payload &&
    (config.method !== 'GET' && config.method !== 'DELETE')
  ) {
    errors.push('payload');
  }

  if (errors.length) {
    throw new Error(`Error! You must pass \`${errors.join('`, `')}\``);
  }
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...config.headers
  };

  const params = {
    url,
    headers,
    method: config.method
  };

  if (params.method !== 'GET') {
    params.data = JSON.stringify(config.payload);
  }
  return axios(params);
}
