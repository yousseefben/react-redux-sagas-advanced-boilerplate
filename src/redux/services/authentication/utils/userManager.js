import { createUserManager } from 'redux-oidc';

const userManagerConfig = {
  client_id: 'example-app',
  redirect_uri: `${window.location.protocol}//${window.location.hostname}${
    window.location.port ? `:${window.location.port}` : ''
  }/callback`,
  response_type: 'token id_token',
  // scope: 'openid profile https://www.googleapis.com/auth/youtube.readonly',
  authority: 'http://localhost:8080/auth/realms/test',
  silent_redirect_uri: `${window.location.protocol}//${
    window.location.hostname
  }${window.location.port ? `:${window.location.port}` : ''}/silent_renew.html`,
  automaticSilentRenew: true,
  filterProtocolClaims: true,
  loadUserInfo: true
};

const userManager = createUserManager(userManagerConfig);

export default userManager;
