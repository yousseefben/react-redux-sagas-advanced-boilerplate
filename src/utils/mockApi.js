const axios = require('axios');
const MockAdapter = require('axios-mock-adapter');

export const mock = new MockAdapter(axios);
