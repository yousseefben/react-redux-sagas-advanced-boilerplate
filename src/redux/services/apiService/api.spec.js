import { request } from './api';
import { mock } from '../../../utils/mockApi';

beforeEach(() => {
  mock.resetHandlers();
});

describe('api', () => {
  describe('request', () => {
    it('should fail without param', () => {
      expect(() => request()).toThrow('Error! You must pass `url`');
    });

    it('should fail to a POST without payload', () => {
      expect(() =>
        request('http://example.com/token', { method: 'POST' })
      ).toThrow('Error! You must pass `payload`');
    });

    it('should execute a GET successfully', done => {
      mock.onAny().reply(200, { hello: 'world' });

      request('http://example.com/token').then(response => {
        expect(response.data).toMatchSnapshot();
        done();
      });
    });
    it('should execute a GET successfully and return response text', done => {
      mock.onAny().reply(200, 'hello');

      request('http://example.com/token').then(response => {
        expect(response.data).toMatchSnapshot();
        done();
      });
    });

    it('should execute a POST successfully', done => {
      mock.onAny().reply(200, { a: 1 });

      request('http://example.com/token', {
        method: 'POST',
        payload: { a: 1 }
      }).then(response => {
        expect(response.data).toMatchSnapshot();
        done();
      });
    });

    it('should reject for a  bad request', done => {
      mock.onAny().reply(400);

      request('http://example.com/token').catch(error => {
        expect(error).toMatchSnapshot();
        done();
      });
    });

    it('should reject for a server error with JSON response', done => {
      mock.onAny().reply(500);
      request('http://example.com/token').catch(error => {
        expect(error).toMatchSnapshot();
        done();
      });
    });
  });
});
