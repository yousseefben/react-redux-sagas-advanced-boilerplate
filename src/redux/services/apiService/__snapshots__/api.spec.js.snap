// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api request should execute a GET successfully 1`] = `
Object {
  "hello": "world",
}
`;

exports[`api request should execute a GET successfully and return response text 1`] = `"hello"`;

exports[`api request should execute a POST successfully 1`] = `
Object {
  "a": 1,
}
`;

exports[`api request should reject for a  bad request 1`] = `[Error: Request failed with status code 400]`;

exports[`api request should reject for a server error with JSON response 1`] = `[Error: Request failed with status code 500]`;
